### PR TITLE
kafka(ticdc): support columns dispatcher.

### LINKED ISSUE #9862: kafka sink support columns dispatcher to calculate the partition number
Column dispatcher allows the user to specify columns to calculate the partition number for each event.

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve?

Issue Number: close #9862 

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
kafka sink support columns dispatcher to allow user to specify columns to calculate the partition number
```

### CODE REVIEW

File: cdc/sink/dmlsink/mq/dispatcher/event_router.go [hunk 1]
```diff
@@ -130,12 +132,24 @@ func (s *EventRouter) GetPartitionForRowChange(
...
-			_, _, ok = table.IndexByName(v.IndexName)
-			if !ok {
+		switch v := partitionDispatcher.(type) {
+		case *partition.IndexValueDispatcher:
+			index := table.GetIndex(v.IndexName)
+			if index == nil {
 				return cerror.ErrDispatcherFailed.GenWithStack(
 					"index not found when verify the table, table: %v, index: %s", table.TableName, v.IndexName)
 			}
+			if !index.Unique {
```
Comment (nongfushanquan) [inline cdc/sink/dmlsink/mq/dispatcher/event_router.go:142]: please add comments about the reason of reporting the error

File: cdc/sink/dmlsink/mq/dispatcher/partition/columns_test.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+				Value: 33,
+			},
+		},
+	}
+
+	p := NewColumnsDispatcher([]string{"col-2", "col-not-found"})
+	_, _, err := p.DispatchRowChangedEvent(event, 16)
+	require.ErrorIs(t, err, errors.ErrDispatcherFailed)
+
+	p = NewColumnsDispatcher([]string{"col2", "col1"})
```
Comment (nongfushanquan) [inline cdc/sink/dmlsink/mq/dispatcher/partition/columns_test.go:77]: how about the case such as NewColumnsDispatcher([]string{"col2", "col2"})

Comment (3AceShowHand) [reply cdc/sink/dmlsink/mq/dispatcher/partition/columns_test.go:77]: col2 will be used twice to calculate the checksum, it's legal and allowed.

**Review by nongfushanquan — APPROVED** [review]

**Review by asddongmen — APPROVED** [review]

File: cdc/sink/dmlsink/mq/dispatcher/partition/columns.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	return &ColumnsDispatcher{
+		hasher:  hash.NewPositionInertia(),
+		Columns: columns,
+	}
+}
+
+// DispatchRowChangedEvent returns the target partition to which
+// a row changed event should be dispatched.
+func (r *ColumnsDispatcher) DispatchRowChangedEvent(row *model.RowChangedEvent, partitionNum int32) (int32, string, error) {
+	r.lock.Lock()
```
Comment (CharlesCheung96) [inline cdc/sink/dmlsink/mq/dispatcher/partition/columns.go:47]: Why a lock is needed here?

File: cdc/sink/dmlsink/mq/dispatcher/partition/columns.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+func NewColumnsDispatcher(columns []string) *ColumnsDispatcher {
+	return &ColumnsDispatcher{
+		hasher:  hash.NewPositionInertia(),
+		Columns: columns,
+	}
+}
+
+// DispatchRowChangedEvent returns the target partition to which
+// a row changed event should be dispatched.
+func (r *ColumnsDispatcher) DispatchRowChangedEvent(row *model.RowChangedEvent, partitionNum int32) (int32, string, error) {
```
Comment (CharlesCheung96) [inline cdc/sink/dmlsink/mq/dispatcher/partition/columns.go:46]:
```suggestion
func (r *ColumnsDispatcher) DispatchRowChangedEvent(row *model.RowChangedEvent, totalPartitionNum int32) (int32, string, error) {
```

**Pull request merged**